8.0 4801 transsmart analysis

daramousk opened the pull request:
@NL66278 Two things here:

1) The prebooking functionality has to be removed
2) Anything that took place on `sale.order` has been removed.

Commit:
[ADD] Port to v2 of the transsmart API, total refactor.

daramousk commented:
@NL66278 This has not been tested yet because I will port this to v2 and that is were the tests will take place. Basically I am concerned with you reviewing the parts of code that I have removed if they are correct and if there is something else to remove.

NL66278 reviewed (dismissed):
:+1:  LGTM

daramousk commented:
@NL66278 More commits added that remove extra clutter. Now I will start porting this to v2

NL66278 commented on delivery_transsmart/__scripts__/res_partner.py:
```diff
@@ -0,0 +1,6 @@
+# -*- coding: utf-8 -*-
+# Copyright 2018 Therp BV <http://therp.nl>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+# TODO: go through all the res.partners. if they do not have a value in those
```
Instead of a script you might also make this part of a migration script, or have it run in an init hook. So will be fully automated.

daramousk commented:
@NL66278 This is a WIP, I will notify when to review soon.

daramousk commented:
@NL66278 I need another review here before I start writing migration scripts, this module contains the whole sprint7. Let me know when I can have this. 

CC: @lfreeke

daramousk commented on delivery_transsmart/wizards/stock_transfer_details.py:
```diff
@@ -0,0 +1,16 @@
+    def do_detailed_transfer(self):
+        self.ensure_one()
+        if self.picking_id._send_to_transsmart():
+            # TODO how will partial moves be taken care off?
```
@lfreeke Maybe this is a question for you? When a picking is to be sent on transsmart, should the user be prohibited or making it a partial or should I create a document for every part of the picking?

NL66278 commented on delivery_transsmart/models/product_template.py:
```diff
@@ -0,0 +1,29 @@
+        string='Service Level Time',
+    )
+    code = fields.Char()
+    nr = fields.Integer('Identifier')
```
maybe rename these kind of fields to transsmart_id? The name nr is not very descriptive...

daramousk replied:
@NL66278 I purposefully let them have the same names as in the Transsmart's API so that I can avoid adding and maintaining extra complexity of mapping around local fields with remote ones. So, the synchronization is simpler as well. These fields are not shown or modifiable by the end user.

NL66278 commented:
@daramousk What I am missing is some README.rst to explain the general concepts and comments in the coding making clear what it does. For instance you add a reference to product.template to res.partner with the name package_type_id. What is this? Why is it on partner? Ofcourse I could grab the transsmart API documentation and extensively study that, but some explanation in the code would really help.

daramousk commented:
@NL66278 I had mistakenly force pushed an empty commit, see now for a full readme.

Commit:
[FIX] Make the errors more specific.

Commit:
[FIX] Remove package_type_id field.

Commit:
[FIX] [no task] Various fixes.

Commit:
[FIX] [no task] Fix transsmart packages missing values.

Commit:
[FIX] [no task] Just show the errors as they come from Transsmart

Commit:
[FIX] Fix the quantity of packages to one.

Commit:
[FIX] Add proper incoterms

Commit:
[FIX] [no task] Add some extra fields on the delivery line

Commit:
[FIX] Make sure that you use the category hs code on a product if the product hs code does not exist

Commit:
[FIX] Ignore state field

Commit:
[FIX] [no task] Remove unessesary field

Commit:
[ADD] [no task] Add booking profiles and the supporting functionality

Commit:
[DEL] [no task] Set service as a selection field and by default non-docs

Commit:
[FIX] [no task] Refactoring and creation of testing modules

Commit:
[FIX] Change persmissions, only admin should be able to change synchronization records

Commit:
[FIX] [no task] Make sync function not protected, change cron job

Commit:
[FIX] Fix issue with non existing field

Commit:
[IMP] Adds a service_level_time field on the sale order. When a stock_picking is created from a sale order then the default value of the service_level_time field that exists on the stock picking will be equal to the one from sale order

NL66278 committed:
[FIX] delivery_transsmart. Should not depend on customer modules.

NL66278 committed:
[IMP] delivery_transsmart. Better named variables and streamlined code.

NL66278 committed:
[IMP] delivery_transsmart. No dependency on web_readonly_by_pass.

NL66278 committed:
[IMP] #4801 Improve compatibility with existing code.

NL66278 committed:
[FIX] delivery_transsmart. Prevent clash with delivery.carrier module.

NL66278 committed:
[IMP] #4801 Support empty phone.

NL66278 committed:
[IMP] delivery_transsmart. Do not send empty keys to Transsmart.

NL66278 committed:
[IMP] delivery_transsmart. Add back request logging.

NL66278 committed:
[IMP] delivery_transsmart. Integrated http rest API.

NL66278 committed:
[FIX] delivery_transsmart. Restore cost_center on sale_order.

NL66278 committed:
[FIX] delivery_transssmart*. Restructured.

NL66278 committed:
[FIX] delivery_transsmart. Correct moment of rate retrieval.

When creating a stock picking from a sale, the code would try to get the rates for the new stockpicking,
but only after that the service level time and cost center would be written on the picking.

NL66278 committed:
[FIX] #4801 Remove unneeded transsmart_confirmed field.

NL66278 committed:
[FIX] #4801 Set correct action name to check connection.

NL66278 committed:
[FIX] #4801 Use field code instead of transsmart_code.

Otherwise there would be problems with existing sale check rules.

NL66278 committed:
[FIX] #4801 Do not copy instance user on connection.

NL66278 committed:
[IMP] #4801 Correct use of prebooking.

NL66278 committed:
[IMP] #4801 Complete 'Get Rates' and 'Prebooking' functionality.

'Get Rates' on stock.picking form is alvailable for sysadmin for debugging
purposes. It will retrieve all rates available for address and package.

'Prebooking' will invoke routing rules to select carrier and update service level time.

When sending package to Transsmart, Prebooking will automatically be invoked if not done before.

NL66278 committed:
[FIX] delivery_transsmart_address_consolidation. Update.

Change this module to be compatible with changes in delivery_transsmart.

NL66278 committed:
[IMP] delivery_transsmart. Fill shipment line description.

Also prevent pickupDate before current date.
Also do some refactoring.
